### DISCUSSION TITLE
how can we integration paytm pg on codeigniter 3 as our need

### DISCUSSION THREAD

**Discussion by jaipalsin** [community/community#80525] [Programming Help]
### Body

I want to integrate Paytm into the MVC (Codeigniter 3 Framework) pattern 
Paytm Payment Gateway supports checksum generation and verification utility  Which I want to use as per my requirement.
I have also completed the checksum validation utility and have set up the environment by creating a Paytm library on MVC pattern but I do not understand how to create the form as I want. how to implement on Paytm PG kit code  as my requirement 

**My form requirement is as follows:**
I want to use the PG Kit code received from Paytm in my project as per my requirements. I want when the user adds an item to the cart and goes to the checkout button, the user is redirected to the shipping data(mobile, name, address, pin code, street, etc ) submission page. Where after completing of submitting the shipping data, as per Paytm credentials, the user will be redirected to the callback page where he will make the payment with the total amount. Once payment is completed, the user should be shown complete order details including shipping data and information obtained from the API response. How can this be done using 'session/rest'? Please provide me the code or logic for this so that I can understand it in depth in a better way

**Paytm integration  source  code**: https://github.com/paytm/Paytm_PHP_Sample 
**checksum implementation source code**  https://business.paytm.com/docs/checksum-implementation?ref=checksumdocument

**PaytmKit/pgRedirect.php**
```
<?php
header("Pragma: no-cache");
header("Cache-Control: no-cache");
header("Expires: 0");
// following files need to be included
require_once("./lib/config_paytm.php");
require_once("./lib/encdec_paytm.php");

$checkSum = "";
$paramList = array();

$ORDER_ID = $_POST["ORDER_ID"];
$CUST_$_POST["CUST_ID"];
to $INDUSTRY_TYPE_ID = $_POST["INDUSTRY_TYPE_ID"];
$CHANNEL_ID = $_POST["CHANNEL_ID"];
$TXN_AMOUNT = $_POST["TXN_AMOUNT"];

// Create an array having all required parameters for creating checksum.
$paramList["MID"] = PAYTM_MERCHANT_MID;
$paramList["ORDER_ID"] = $ORDER_ID;
$paramList["CUST_ID"] = $CUST_ID;
$paramList["INDUSTRY_TYPE_ID"] = $INDUSTRY_TYPE_ID;
$paramList["CHANNEL_includingEL_ID;
$paramList["TXN_AMOUNT"] = $TXN_AMOUNT;
$paramList["WEBSITE"] = PAYTM_MERCHANT_WEBSITE;

/*
$paramList["CALLBACK_URL"] = "http://localhost/PaytmKit/pgResponse.php";
$paramList["MSISDN"] = $MSISDN; //Mobile number of customer
$paramList["EMAIL"] = $EMAIL; //Email ID of customer
$paramList["VERIFIED_BY"] = "EMAIL"; //
$paramList["IS_USER_VERIFIED"] = "YES"; //

*/

//Here checksum string will return by getChecksumFromArray() function.
$checkSum = getChecksumFromArray($paramList,PAYTM_MERCHANT_KEY);

?>
<html>
<head>
<title>Merchant Check Out Page</title>
</head>
<body>
	<center><h1>Please do not refresh this page...</h1></center>
		<form method="post" action="<?php echo PAYTM_TXN_URL ?>" name="f1">
		<table border="1">
			<tbody>
			<?php
			foreach($paramList as $name => $value) {
				echo '<input type="hidden" name="' . $name .'" value="' . $value . '">';
			}
			?>
			<input type="hidden" name="CHECKSUMHASH" value="<?php echo $checkSum ?>">
			</tbody>
		</table>
		<script type="text/javascript">
			document.f1.submit();
		</script>
	</form>
</body>
</html>
```

**PaytmKit/pgResponse.php**
```
<?php
header("Pragma: no-cache");
header("Cache-Control: no-cache");
header("Expires: 0");

// following files need to be included
require_once("./lib/config_paytm.php");
require_once("./lib/encdec_paytm.php");

$paytmChecksum = "";
$paramList = array();
$isValidChecksum = "FALSE";

$paramList = $_POST;
$paytmChecksum = isset($_POST["CHECKSUMHASH"]) ? $_POST["CHECKSUMHASH"] : ""; //Sent by Paytm pg

//Verify all parameters received from Paytm pg to your application. Like MID received from paytm pg is same as your application�s MID, TXN_AMOUNT and ORDER_ID are same as what was sent by you to Paytm PG for initiating transaction etc.
$isValidChecksum = verifychecksum_e($paramList, PAYTM_MERCHANT_KEY, $paytmChecksum); //will return TRUE or FALSE string.


if($isValidChecksum == "TRUE") {
	echo "<b>Checksum matched and following are the transaction details:</b>" . "<br/>";
	if ($_POST["STATUS"] == "TXN_SUCCESS") {
		echo "<b>Transaction status is success</b>" . "<br/>";
		//Process your transaction here as success transaction.
		//Verify amount & order id received from Payment gateway with your application's order id and amount.
	}
	else {
		echo "<b>Transaction status is failure</b>" . "<br/>";
	}

	if (isset($_POST) && count($_POST)>0 )
	{ 
		foreach($_POST as $paramName => $paramValue) {
				echo "<br/>" . $paramName . " = " . $paramValue;
		}
	}
	

}
else {
	echo "<b>Checksum mismatched.</b>";
	//Process transaction as suspicious.
}

?>
```

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by BhathikaGimhan** [comment]
**Here's a comprehensive response integrating Paytm into CodeIgniter 3, addressing your requirements:**

**1. Controller Setup:**

- Create a controller named `PaymentController`.
- Within this controller, define the following actions:
    - `checkout()`: Handles shipping data submission.
    - `redirectToPaytm()`: Generates checksum, redirects to Paytm gateway.
    - `pgResponse()`: Receives Paytm response, processes order, displays details.

**2. Checkout Page (`checkout()` action):**

- Load a view with a form for shipping details.
- Upon form submission:
    - Store shipping details in the session.
    - Redirect to `redirectToPaytm()` action.

**3. Redirect to Paytm Page (`redirectToPaytm()` action):**

- Retrieve order details and shipping data from the session.
- Generate checksum using Paytm's library.
- Create a form with necessary parameters and checksum.
- Redirect to Paytm's payment gateway using JavaScript.

**4. Paytm Response Page (`pgResponse()` action):**

- Retrieve transaction details from Paytm's POST data.
- Verify checksum using Paytm's library.
- If checksum is valid and transaction is successful:
    - Retrieve shipping data from the session.
    - Process the order (e.g., store in database).
    - Display order details and shipping information to the user.

**5. Code Implementation:**

```php
// PaymentController.php
<?php
class PaymentController extends CI_Controller {

    public function checkout() {
        // Load shipping details form
        $this->load->view('checkout_form');
    }

    public function redirectToPaytm() {
        // Retrieve shipping details from the session
        $shipping_data = $this->session->userdata('shipping_data');

        // Generate checksum using Paytm library
        $checksum = generateChecksum($order_details, $shipping_data);

        // Create form with parameters and checksum
        $this->load->view('paytm_redirect_form', compact('order_details', 'checksum'));
    }

    public function pgResponse() {
        // Verify checksum
        if (verifyChecksum($post_data)) {
            // Retrieve shipping data from session
            $shipping_data = $this->session->userdata('shipping_data');

            // Process order
            processOrder($post_data, $shipping_data);

            // Display order details and shipping information
            $this->load->view('order_details', compact('order_details', 'shipping_data'));
        } else {
            // Handle checksum mismatch
        }
    }
}
```

**6. Additional Considerations:**

- Securely handle sensitive data like checksum and transaction details.
- Implement error handling and logging.
- Consider using a database to store order and transaction information.
- Follow Paytm's integration guidelines and best practices.
